Type checking for template expressions

ktsn opened the pull request:
This is work in progress PR which enables Vetur to type check template expressions. 

The approach is the same as I wrote in #209 - transforming template html to TypeScript AST, then type checking it. I use [`vue-eslint-parser`](https://github.com/mysticatea/vue-eslint-parser) to transform HTML string to AST as it considers vue specific things like directives. I wrote the transformer that converts ESLint AST to TypeScript AST. Note that the output TypeScript AST does not make sense in runtime to simplify the implementation as it is meant to be used for extracting type info from template expression.

With this feature, we may also able to provide completion in template and component props type checking but I would like to leave them for another PR since this PR is already large without them 😅 

BTW, should we disable this feature in default and let the users enable it by configuration?

## TODO for this PR

- [x] A overall good structure
- [x] Good test coverage
- [x] Update diagnostics when script block is updated
- [x] Add a new option `vetur.experimental.templateTypeCheck`

## Remaining Tasks (after this PR)

- [x] Check other directives expression. (currently only works in `v-bind`/`v-on`)
- Support filter
- Support scoped slot
- [x] Suppress `An object literal cannot have multiple properties with the same name in strict mode.` error.
  - e.g. `<div class="foo" :class="bar"></div>` provides this error.
- Support type narrowing by `v-if`
  - https://next.angular.io/guide/aot-compiler#type-narrowing

fix #209

ktsn commented:
😄 
![template-type-checking](https://user-images.githubusercontent.com/2194624/35811478-efc02334-0ad1-11e8-8c91-3ffff62d6fd3.gif)

ktsn commented:
It probably be better to see diff via https://github.com/vuejs/vetur/pull/681/files?w=1 when looking into `serviceHost.ts` since there is an indentation change.

HerringtonDarkholme commented on server/src/modes/script/preprocess.ts:
```diff
@@ -1,65 +1,111 @@
 import * as ts from 'typescript';
 import * as path from 'path';
+import { parse } from 'vue-eslint-parser';
```
We can implement parsing in VLS at all, so no additional dependency/script is needed. Actually template completion needs it too.

But I don't have time to implement Vue specific elements 😞 . For now eslint-parser is the only option.

ktsn replied:
Yes. I actually tried to extend the parser in VLS on the first time but I ended up using vue-eslint-parser because I would like to focus finishing essential implementation of template type checking at first.
In that case, I can work on the VSL template parser after this PR is finished 🙂

HerringtonDarkholme commented on server/src/modes/script/transformTemplate.ts:
```diff
@@ -0,0 +1,457 @@
+import * as ts from 'typescript';
+import { AST } from 'vue-eslint-parser';
+
+export const renderHelperName = '__veturRenderHelper';
```
Renaming it to `__vlsRenderHelper` is better since vue-language-server is designed to be client agnostic.

HerringtonDarkholme commented:
Also https://next.angular.io/guide/aot-compiler#phase-3-binding-expression-validation is a good reference.

HerringtonDarkholme commented:
Road Test list:

https://github.com/gitlabhq/gitlabhq
https://github.com/bootstrap-vue/bootstrap-vue
https://github.com/ElemeFE/element/
https://github.com/algolia/vue-instantsearch
https://github.com/vuetifyjs/vuetify
https://github.com/samiheikki/javascript-guessing-game
https://github.com/phanan/koel
https://bitbucket.org/sellvana/core/src/4231c995aff9785ca60a550b0ee06c286f4bf619?at=sn1
https://github.com/search?utf8=%E2%9C%93&q=language%3AVue+stars%3A%3E1000&type=Repositories&ref=advsearch&l=Vue&l=

octref commented:
First -- great work! Thanks for pushing this forward! 🎉 

> BTW, should we disable this feature in default and let the users enable it by configuration?

Yes. We should put an option like `vetur.experimental.templateTypeCheck` before it.

But I think currently the biggest problem is that there is no longer a clean separation of mode, as the type checking exists on both template/script modes. If we want to do suggestion/hovering and other LS features, it makes sense to create a `templateInterpolation` mode.

> TODO

Let's aim for a MVP, where we have:
- A overall good structure
- Good test coverage
- Update diagnostics when script block is updated

Can you add me to your fork so I can push changes? I'll have some time to look deeper into this PR this weekend. Before that I'll spend some time looking into some other smaller, long-standing bugs...

@HerringtonDarkholme Yeah I agree we should start adding some integration tests running VLS against real-world repos. This is one of the things I plan to look into this month.

ktsn commented:
@octref @HerringtonDarkholme I just added you as collaborators of my fork. Also I've updated the TODO list.

> But I think currently the biggest problem is that there is no longer a clean separation of mode, as the type checking exists on both template/script modes. If we want to do suggestion/hovering and other LS features, it makes sense to create a templateInterpolation mode.

Yeah, that is I'm concerning but I didn't find a clean way to implement this. Do you mean creating a new mode `templateInterpolation` and separate it with template/script mode?

HerringtonDarkholme commented:
I'm working on a branch to parse interpolation in VLS. 

https://github.com/HerringtonDarkholme/vetur/tree/template

yringler commented:
Very cool. Would this work with[ vue-class-component ](https://github.com/vuejs/vue-class-component)and [vue-property-decorator](https://github.com/kaorun343/vue-property-decorator)?

yringler commented:
I compiled from your branch, and... this works with vue-class-component and freinds! So cool. I would love to help out with this.
In the meantime, perhaps this branch should be available in the vs code extension repository, as a beta or something? because this is super useful.

yringler commented:
RE the "An object literal cannot have multiple properties with the same name in strict mode.":
@ktsn , would you happen to know of a way to suppress the error? A setting somewhere?

githubmann commented:
hey, guy, any progress?

NickBolles commented:
@ktsn How's this going? is there anything that needs work that I could pitch in on?

hmexx commented:
I'd love to try this feature. Any chance we can see a beta of it in an official vetur release?

octref commented:
I'm making some changes in Vetur to accommodate this PR and #1083. We'll hopefully ship completion on February and this in March. Please just stay patient.

yringler commented:
In the off chance that this isn't a known issue:
<img src="https://user-images.githubusercontent.com/4070825/54615180-07bb9700-4a34-11e9-92f8-eac7096e4183.png" alt="image" width="3840">
Note the error: ", paylo doesn't exist", when the variable is actually _payload_, which does exist, Seems there's a character shift going on somewhere, and it's looking for a property with the name ", paylo".

octref commented on test/lsp/templateDiagnostics/basic.test.ts:
```diff
@@ -0,0 +1,96 @@
+    await sleep(FILE_LOAD_SLEEP_TIME);
+  });
+
+  const tests: TemplateDiagnosticTest[] = [
```
Integration tests are better. Although they run slower they share the setup that user have. Unit testing just the transformer part is good, but for asserting diagnostics let's use this.

octref commented on server/src/modes/script/transformTemplate.ts:
```diff
@@ -0,0 +1,519 @@
+        vBind
+      );
+    }
+    // Todo: What about VExpressionContainer?
```
This seems to have changed to be `VExpressionContainer | VIdentifier`. How do we do transform in this case?

octref commented on server/src/modes/script/transformTemplate.ts:
```diff
@@ -0,0 +1,519 @@
+      // e.g. @click="onClick"
+      return setTextRange(ts.createPropertyAssignment(setTextRange(ts.createIdentifier(name.name), vOn.key), exp), vOn);
+    }
+    // Todo: What about VExpressionContainer?
```
This seems to have changed to be `VExpressionContainer | VIdentifier`. How do we do transform in this case?

octref commented on server/src/modes/script/transformTemplate.ts:
```diff
@@ -0,0 +1,519 @@
+ */
+function isSpecialAttribute(node: AST.VAttribute | AST.VDirective): boolean {
+  // Todo: Why check node.key.arguments here?
+  // const name = isVAttribute(node) ? node.key.name : node.key.argument;
```
@ktsn Here's your original code. Why are we checking against `node.key.arguments`?

ktsn replied:
Because when the `node` is `v-bind` directive, the HTML attribute name is placed on `node.key.argument`. I should've leave some comment for the intent.

ktsn replied:
I ended up removing this function because the motivation was to avoid duplicated property name error when both `class` and `:class` is used on the same tag. Simply checking `class` and `style`, and skip them would be enough.

octref commented:
@ktsn 

I ticked most of the boxes for you. Because the diff between your PR and current master is too large I created https://github.com/vuejs/vetur/tree/ktsn-types (the diff there might be easier to consume than the ones in this PR) and merged that into your PR. I also left a bunch of comments explaining various changes I did. Take a look and let me know anything you don't understand.

If you don't have enough time, I can also try to finish this myself. The remaining pieces are:

- Fix the todos in `transformTemplate`. Especially the ones concerning `VExpressionContainer | VIdentifier`
- Fix the crash on `v-on.vue`.

There are a few unrelated integration tests failing. Ignore them and I can fix them.

octref commented:
I fixed some tests related to upgrading Vue types.
 There are 2 more failing tests which I believe is caused by TS Server trying to resolve `.vue.template`

octref commented:
I see, the problem is you try to set zero pos for all the template-script code, so TS resolved them to invalid locations and throws there.

I think we can try maybe appending these lines to the end of file and map them back? This way we can get other features such as hover, jump to definition (from template), etc working as well. I'll give it a try.

ktsn commented on server/src/modes/script/preprocess.ts:
```diff
@@ -1,64 +1,112 @@
+
+    if (isVirtualVueTemplateFile(fileName)) {
+      // TODO: share the logic of transforming the code into AST
+      // with the template mode
```
To use Vetur's template parser instead of vue-eslint-parser. I guess I'll work on it after this PR.

ktsn commented:
@octref I'm trying to execute e2e test to test template diagnostics on my local but it fails immediately for some reason. The entire log is the below:

<details>
<pre>$ yarn test:e2e
yarn run v1.13.0
$ node ./dist/test/codeTestRunner.js
### Vetur Integration Test ###

Test folder: dist/test/grammar
Workspace:   test/grammar
[vscode-icons] v8.5.0 activated!





  colorization

    1) should colorize Functional.vue right

    2) should colorize Html.vue right

    3) should colorize Pug.vue right





  0 passing (6ms)

  3 failing



  1) colorization
       should colorize Functional.vue right:
     Error: No TM Grammar registered for this language.
      at t._createGrammar (file://.vscode-test/stable/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:4451:30)
      at t.createGrammar (file://.vscode-test/stable/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:4450:794)
      at e.captureSyntaxTokens (file://.vscode-test/stable/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:4443:784)
      at file://.vscode-test/stable/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:4444:316
      at <anonymous>


  2) colorization
       should colorize Html.vue right:
     Error: No TM Grammar registered for this language.
      at t._createGrammar (file://.vscode-test/stable/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:4451:30)
      at t.createGrammar (file://.vscode-test/stable/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:4450:794)
      at e.captureSyntaxTokens (file://.vscode-test/stable/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:4443:784)
      at file://.vscode-test/stable/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:4444:316
      at <anonymous>


  3) colorization
       should colorize Pug.vue right:
     Error: No TM Grammar registered for this language.
      at t._createGrammar (file://.vscode-test/stable/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:4451:30)
      at t.createGrammar (file://.vscode-test/stable/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:4450:794)
      at e.captureSyntaxTokens (file://.vscode-test/stable/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:4443:784)
      at file://.vscode-test/stable/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:4444:316
      at <anonymous>






Exit code:   1
Done

✨  Done in 11.78s.
</pre></details>

I did the steps that stated in the contribution guide (install and compile). The error also happens when I on the current master branch. Installed VSCode version is 1.33.1.
Do you have any idea to fix this issue?

ktsn commented:
Never mind about the above comment. If I quit all VSCode process, the test works.

ktsn commented on server/src/modes/script/javascript.ts:
```diff
@@ -102,35 +102,73 @@ export async function getJavascriptMode(
-        return [];
+      const templateDiags = getTemplateDiagnostics();
+      const scriptDiags = getScriptDiagnostics();
+      return [...templateDiags, ...scriptDiags];
```
@octref I realized this might cause an edge case. When we do not have `<script>` block in SFC, template diagnostic won't work. I'm not sure how we should include script language mode in that case 🤔

octref replied:
@ktsn When you don't have `<script>` block, what template diagnostic do you want?

ktsn replied:
Mainly detecting some typos. For example, sometimes I have some typo in template expression like:

```vue
<template>
  <ChildComp @click="$('click')" /> <!-- <= typo of $emit('click') -->
</template>
```

Another example is when we use scoped slot component (scoped slot variables not handled currently though):

```vue
<template>
  <apollo-query :query="..." v-slot="{ isLoading, result }">
    <-- typo of v-if="isLoading" -->
    <div v-if="loading">Loading...</div>

    <div v-else-if="result.data">{{ result.data }}</div>
  </apollo-query>
</template>
```

octref replied:
I'll open a new issue for this.

Let's keep these two things separate:

- Where errors are generated — Should be in JS mode
- When should diagnostics be re-generated — For `<template>` part, if there's a change in template interpolation, it should re-trigger JS mode diagnostics

ktsn commented:
> - Fix the todos in transformTemplate. Especially the ones concerning VExpressionContainer | VIdentifier
> - Fix the crash on v-on.vue.

Fixed them.

I also tweaked template transformer a bit to avoid false-positive diagnostics about `v-slot` value https://github.com/vuejs/vetur/pull/681/commits/2923e08a3ea8ecb2b2f2c68da75be1b20ec13557. I'll work on it a later PR.

octref commented:
> Never mind about the above comment. If I quit all VSCode process, the test works.

Yeah, if you are testing against VS Code stable, having another VS Code stable open would cause issues. I should have surfaced the error though.

octref commented:
Merged. @ktsn Congratulations 🎉 👍 

Do you mind creating a new issue to track what you plan to do? I can add to that issue my side of things.

From my perspective I'd like to see if I can make completion work from the `.template` files, and I'd want to implement a mapping between the `.template` files and the Vue templates in SFC so we get hover, jump to definition, find references etc working.

ktsn commented:
@octref Finally 🎉  Thank you for cooperating on this!
Sure, I'll create new issues for remaining tasks.

> hover, jump to definition, find references etc working

Yeah, they are also promising features. I'm so excited we will have them 😄

mpvosseller commented:
This is great! Is there any way to integrate this work into command line builds to get the same errors/warnings there?

octref commented:
@mpvosseller I updated https://github.com/vuejs/vetur/issues/1149, you can subscribe to it.

trusktr commented:
Hello I'm interested in making my own template system. F.e. I'd like to use `@if=""` instead of `v-if=""`, or something. What would be involved in order to make this work with my own syntax?

octref commented:
There are some more details in http://blog.matsu.io/generic-vue-template-interpolation-language-features.
